components/VM_Arm/src/modules/virtio_con.c

```diff
@@ -22,61 +42,160 @@
+static serial_conn_t connections[NUM_PORTS];
+
+/* defined in virtio_console_emul.c */
+void virtio_console_putchar(int port, virtio_emul_t *con, char *buffer, uint32_t head, uint32_t tail) WEAK;
+
+/**
+ * Writes data to the guest console attached to a port, and sets the head of the receive
+ * ring buffer to the right place.
+ *
+ * @param port Port number
+ */
+static void virtio_con_notify_recv(int port)
+{
+    volatile serial_shmem_t *buffer = connections[port].recv_buf;
+    virtio_console_putchar(port, virtio_con->emul, (char *)buffer->buf, buffer->head, buffer->tail);
+    buffer->head = buffer->tail;
+}
 
+/**
+ * Callback function for camkes virtqueue notification. Polls each receive ring buffer
+ * to see if there are data pending.
+ *
+ * @see main.c (struct device_notify) for parameters
+ */
 static int handle_serial_console(vm_t *vmm, void *cookie UNUSED)
 {
-    struct {
-        uint32_t head;
-        uint32_t tail;
-        char buf[BUFSIZE];
-    } volatile *buffer = serial_getchar_buf;
-
-    int count = 0;
-    while (buffer->head != buffer->tail) {
-        txbuf[count % BUFSIZE] = buffer->buf[buffer->head];
-        buffer->head = (buffer->head + 1) % BUFSIZE;
-        count += 1 % BUFSIZE;
+    /* Poll each ring buffer to see if data was added */
+    for (int i = 0; i < NUM_PORTS; i++) {
+        if (connections[i].recv_buf->head != connections[i].recv_buf->tail) {
+            virtio_con_notify_recv(i);
+        }
     }
-    virtio_console_putchar(virtio_con->emul, txbuf, count);
     return 0;
 }
 
-static void emulate_console_putchar(char c)
+/**
+ * Forwards data to the destination port using camkes virtqueue, and notifies the dest VM.
+ *
+ * @param port Port number
+ * @param c Data to forward
+ */
+static void tx_virtcon_forward(int port, char c)
 {
-#if CONFIG_NUM_DOMAINS > 1 && CONFIG_PRINTING
-    seL4_DebugPutChar(c);
-#else
-    batch_buf->buf[batch_buf->tail] = c;
-    batch_buf->tail = (batch_buf->tail + 1) % BUFSIZE;
-    if (c == '\n' || batch_buf->tail + 1 == batch_buf->head) {
-        batch_batch();
+    if (port >= NUM_PORTS) {
+        return;
+    }
+
+    volatile serial_shmem_t *buffer = connections[port].send_buf;
+    buffer->buf[buffer->tail] = c;
+    buffer->tail = (buffer->tail + 1) % BUFSIZE;
+
+    /* If newline or staging area full, it's time to send */
+    if (c == '\n' || (buffer->tail + 1) % BUFSIZE == buffer->head) {
+        connections[port].notify();
     }
-#endif
 }
 
+/* camkes-generated symbols */
+/* regions shared by the SerialServer and the VMM */
+extern serial_shmem_t volatile *batch_buf;
+extern serial_shmem_t volatile *serial_getchar_buf;
+
 extern seL4_Word serial_getchar_notification_badge(void);
+
+/**
+ * Initialize a virtio console backend. This includes:
+ * - adding an IO port handler
+ * - adding a PCI entry
+ * - initializing the emul layer
+ * - setting up connections (includes hvc0)
+ *
+ * @param vm The vm handler of the caller
+ * @param cookie Unused
```

Comment:
I'd still put in some dummy value here to force the caller to use this properly. Otherwise the callers might ignore this, and we have a potential problem in the future when this cookie is really needed. Practically, we could also put `connections` in here, as this seems our global context at the moment.